Implement sequences (arrays, vectors, strings)

ActuallyEnbeon opened the pull request:
Array and vectors are able to be created and indexed, although they don't have any specific methods yet (e.g. length, append, etc) 

Ranges (`[x:y]`) also cannot be compiled

Once we get proper memory management implemented, we can rework as much of this as we need to

ActuallyEnbeon committed:
Update DESIGN.md with array and vector syntax

ActuallyEnbeon committed:
Add array and vector types to parser

Also adds the arr and vec keywords

ActuallyEnbeon committed:
Add array literals to parser

ActuallyEnbeon committed:
Add type checking for arrays and vectors

ActuallyEnbeon committed:
Add array literals and vectors to interpreter

ActuallyEnbeon committed:
Merge branch 'main' into array

ActuallyEnbeon committed:
Update DESIGN.md with further array syntax

ActuallyEnbeon committed:
Added arr[] and vec[] literals

Array and vector literals can now be specifically created. Oddly, assigning an arr[] to a variable of type vec actually works, when it probably shouldn't

ActuallyEnbeon committed:
Adjust frontend for sequence types (#18)

The backend no longer works, need to wait until the IR is done before continuing with it

aurorusidk committed:
Merge branch 'main' into array

ActuallyEnbeon committed:
Implement sequences in the IR

Includes sequence expressions and sequence types, but not index expressions

ActuallyEnbeon committed:
Implement index expressions in IR

ActuallyEnbeon committed:
Implement sequences and indexes in type checker

ActuallyEnbeon committed:
Implement sequences and indexes in interpreter

Includes ranges [x:y] but not sequence methods (length, append, etc.)

aurorusidk committed:
Make builtins use the correct type ref

aurorusidk committed:
Clean up dead code for arrays

ActuallyEnbeon committed:
Improve type checker and interpreter for sequences

It's now not possible to assign an array to a vector variable or vice versa. Range literals are also now directly indexable

ActuallyEnbeon committed:
Update DESIGN.md to match main

ActuallyEnbeon committed:
Simplify type checker logic for sequences

The logic now operates using IR types, bringing it in line with the rest of the type checker

ActuallyEnbeon committed:
Mark out compiler for sequences

MultipleNewts committed:
Add chars and strings to interpreter #19

Co-Authored-By: Wren <75564966+aurorusidk@users.noreply.github.com>
Co-Authored-By: Enbeon <85409980+ActuallyEnbeon@users.noreply.github.com>

MultipleNewts committed:
Update design docs #19

Co-Authored-By: Wren <75564966+aurorusidk@users.noreply.github.com>
Co-Authored-By: Enbeon <85409980+ActuallyEnbeon@users.noreply.github.com>

aurorusidk committed:
Improve IR type info for strings

aurorusidk committed:
Add compilation for string literals and indexing (#19)

ActuallyEnbeon committed:
Add rough compilation for constant arrays

This code is terrible but at least it works a little bit;
will need to spend more time later making it suck less

ActuallyEnbeon committed:
Add array type separate from string type in compiler

Created a new array type that mostly mirrors the string type.
This is a temporary measure to make arrays work natively while memory management gets up and running.

ActuallyEnbeon committed:
Remove incorrect comments

ActuallyEnbeon committed:
Add full compilation for arrays

Includes support for arrays with nonconstant elements

ActuallyEnbeon committed:
Add compilation for vectors (and unspecified sequences)

There are still issues surrounding directly indexing a sequence literal,
but this code works for the most part

ActuallyEnbeon committed:
Resolved segfault when directly indexing sequence literal

ActuallyEnbeon committed:
Update testing with changes to arrays

Compiler testing is completely borked but the rest works now

ActuallyEnbeon committed:
Merge branch 'main' into array

aurorusidk commented:
I think the sync with main must have broken something.
```
./src/python/builtin.py:19:15: F821 undefined name 'type_defs'
string_type = type_defs.BasicType(BasicTypeKind.STR, BasicTypeFlag.STRING, "str")
              ^
./src/python/builtin.py:35:13: F821 undefined name 'type_defs'
            type_defs.FunctionType(
            ^
./src/python/builtin.py:36:17: F821 undefined name 'type_defs'
                type_defs.ArrayType(types["int"], None),
                ^
./src/python/builtin.py:43:22: F821 undefined name 'type_defs'
                hint=type_defs.ArrayType(types["int"], None),
                     ^
./src/python/builtin.py:44:25: F821 undefined name 'type_defs'
                checked=type_defs.ArrayType(types["int"], None),
                        ^
./src/python/interpreter.py:264:39: F821 undefined name 'builtin'
                elements = [StaObject(builtin.scope.lookup("int"), i) for i in range(start, end)]
                                      ^
./src/python/interpreter.py:266:46: F821 undefined name 'builtin'
                    StaArray(types.ArrayType(builtin.scope.lookup("int"), end-start), elements)
                                             ^
7     F821 undefined name 'type_defs'
```

ActuallyEnbeon committed:
Re-add missing import

An import was removed on main so when main was merged into array the code broke

aurorusidk committed:
Re-add other missing import

ActuallyEnbeon committed:
Merge branch 'main' into array

ActuallyEnbeon committed:
Fix a syntax error

aurorusidk committed:
Merge branch 'main' into array

aurorusidk committed:
Update compiler testing to pass tests

aurorusidk committed:
Format commented compiler tests properly

aurorusidk committed:
Update compilation function used in __main__.py
